symplyphysics/conditions/thermodynamics/internal_energy_is_first_order_homogeneous_function.py

```diff
@@ -0,0 +1,56 @@
+from sympy import Eq
+from symplyphysics import (
+    units,
+    dimensionless,
+    Symbol,
+    Function,
+    print_expression,
+)
+from symplyphysics.core.expr_comparisons import expr_equals
+from symplyphysics.laws.thermodynamics import internal_energy_differential as internal_energy_law
+
+# Description
+## Internal energy is a first-order homogeneous function of its internal variables (entropy,
+## volume, and particle count).
+
+# Law: U(k * S, k * V, k * N) = k * U(S, V, N) forall k
+## U = U(S, V, N) - internal energy as a function of its natural variables
+## S - entropy
+## V - volume
+## N - particle count
+## k - any (real) number
+
+internal_energy = Function("internal_energy", units.energy)
+entropy = Symbol("entropy", units.energy / units.temperature)
+volume = Symbol("volume", units.volume)
+particle_count = Symbol("particle_count", dimensionless)
+factor = Symbol("factor", dimensionless)
+
+homogeneity_condition = Eq(
+    internal_energy(factor * entropy, factor * volume, factor * particle_count),
+    factor * internal_energy(entropy, volume, particle_count),
+)
+
+# Derive from the formula of internal energy differential
+
+_internal_energy_expr = internal_energy_law.law.rhs
+
+_lhs = homogeneity_condition.lhs.subs(
+    internal_energy(factor * entropy, factor * volume, factor * particle_count),
+    _internal_energy_expr.subs({
+        internal_energy_law.entropy_change: factor * internal_energy_law.entropy_change,
+        internal_energy_law.volume_change: factor * internal_energy_law.volume_change,
+        internal_energy_law.particle_count_change: factor * internal_energy_law.particle_count_change,
+    }),
+)
+
+_rhs = homogeneity_condition.rhs.subs(
+    internal_energy(entropy, volume, particle_count),
+    _internal_energy_expr,
+)
+
+assert expr_equals(_lhs, _rhs)
+
+
+def print_law() -> str:
```

Comment:
First of all, maybe we can have a separate function for printing laws that accepts modules as input? It's not that difficult to write and we will thus not need to write this boilerplate in every law.

Also I don't think I understand what counts as condition vs law and what doesn't. Maybe just put everything in laws, just without the calculation function? In my opinion, here we have a condition that characterises the internal energy function and that's pretty much it. It's not really a "law". And I can't think of any sound calculation function. I doubt it will ever be needed.

Reply:
I tend to agree that `print_law` is useless. We do not need printing that accepts module though.
Conditions let us attach some property (condition) to the law. Conditions are generally laws, but they are too general and too simple to add `calculate` to them. If we simply remove `calculate` in law, people will try to fix it, because it looks like a programmer error. While for conditions it is logical not to have `calculate` function.

Reply:
Removing all of them is a major refactoring. We are keeping them for now.

Reply:
> But I don't have much against moving it to laws if you insist.
I do not insist on it. Actually I'm leaning towards keeping it a condition. The problem is, I see it as a more general condition, ie `F(k * x, k * y, ... , k * z) = k * F(x, y, ..., z)`
But I do not see, how to do it for undefined set of parameters.

Reply:
I'm afraid there's no way to make it more general: in the context of thermodynamic potentials this requires four different laws because some of the parameters be intensive and must be omitted during the scaling. Say, for Gibbs energy it's $`G(T, p, kN) = kG(T, p, N)`$.

Reply:
> Removing all of them is a major refactoring. We are keeping them for now.

Do we need to write them in new laws?

Reply:
You can omit them.